search/processor.cpp

```diff
@@ -431,98 +494,40 @@ void Processor::ProcessEmojiIfNeeded(strings::UniString const & token, size_t in
   }
 }
 
-void Processor::SetQuery(string const & query)
+int Processor::GetQueryIndexScale(m2::RectD const & viewport) const
 {
-  m_query = query;
-
-  /// @todo Why Init is separated with SetQuery?
-  ASSERT(m_tokens.empty(), ());
-  ASSERT(m_prefix.empty(), ());
-
-  // Following code splits input query by delimiters except hash tags
-  // first, and then splits result tokens by hashtags. The goal is to
-  // retrieve all tokens that start with a single hashtag and leave
-  // them as is.
-
-  vector<strings::UniString> tokens;
-  {
-    search::DelimitersWithExceptions delims(vector<strings::UniChar>{'#'});
-    SplitUniString(NormalizeAndSimplifyString(query), MakeBackInsertFunctor(tokens), delims);
-  }
-
-  search::Delimiters delims;
-  {
-    buffer_vector<strings::UniString, 32> subTokens;
-    for (auto const & token : tokens)
-    {
-      size_t numHashes = 0;
-      for (; numHashes < token.size() && token[numHashes] == '#'; ++numHashes)
-        ;
-
-      // Splits |token| by hashtags, because all other delimiters are
-      // already removed.
-      subTokens.clear();
-      SplitUniString(token, MakeBackInsertFunctor(subTokens), delims);
-      if (subTokens.empty())
-        continue;
-
-      if (numHashes == 1)
-        m_tokens.push_back(strings::MakeUniString("#") + subTokens[0]);
-      else
-        m_tokens.emplace_back(move(subTokens[0]));
+  return search::GetQueryIndexScale(viewport);
+}
 
-      for (size_t i = 1; i < subTokens.size(); ++i)
-        m_tokens.push_back(move(subTokens[i]));
-    }
-  }
+void Processor::Search(Results & results, size_t limit)
+{
+  if (m_tokens.empty())
+    SuggestStrings(results);
 
-  // Assign prefix with last parsed token.
-  if (!m_tokens.empty() && !delims(strings::LastUniChar(query)))
-  {
-    m_prefix.swap(m_tokens.back());
-    m_tokens.pop_back();
-  }
+  v2::Geocoder::Params params;
+  InitParams(false /* localitySearch */, params);
+  params.m_mode = m_mode;
 
-  int const maxTokensCount = MAX_TOKENS - 1;
-  if (m_tokens.size() > maxTokensCount)
-    m_tokens.resize(maxTokensCount);
+  params.m_pivot = GetPivotRect();
+  params.m_accuratePivotCenter = GetPivotPoint();
+  m_geocoder.SetParams(params);
 
-  // Assign tokens and prefix to scorer.
-  m_keywordsScorer.SetKeywords(m_tokens.data(), m_tokens.size(), m_prefix);
+  m_geocoder.GoEverywhere(m_preRanker);
 
-  // get preffered types to show in results
-  m_prefferedTypes.clear();
-  ForEachCategoryTypes(v2::QuerySliceOnRawStrings<decltype(m_tokens)>(m_tokens, m_prefix),
-                       [&](size_t, uint32_t t)
-                       {
-                         m_prefferedTypes.insert(t);
-                       });
+  FlushResults(params, results, false /* allMWMs */, limit, false /* oldHouseSearch */);
```

Comment:
FlushResults is always called with false for allMWMs, do we need this param too?